obfuscation_test.go

```diff
@@ -0,0 +1,321 @@
+package sphinx
+
+import (
+	"bytes"
+	"encoding/hex"
+	"testing"
+
+	"github.com/roasbeef/btcd/btcec"
+)
+
+// TestOnionFailure checks the ability of sender of payment to decode the
+// obfuscated onion error.
+func TestOnionFailure(t *testing.T) {
+	// Create numHops random sphinx paymentPath.
+	paymentPath := make([]*btcec.PublicKey, 5)
+	for i := 0; i < len(paymentPath); i++ {
+		privKey, err := btcec.NewPrivateKey(btcec.S256())
+		if err != nil {
+			t.Fatalf("unable to generate random key for sphinx node: %v", err)
+		}
+		paymentPath[i] = privKey.PubKey()
+	}
+	sessionKey, _ := btcec.PrivKeyFromBytes(btcec.S256(),
+		bytes.Repeat([]byte{'A'}, 32))
+
+	// Reduce the error path on one node, in order to check that we are
+	// able to receive the error not only from last hop.
+	errorPath := paymentPath[:len(paymentPath)-1]
+
+	failureData := []byte("some kek-error data")
+	sharedSecrets := generateSharedSecrets(paymentPath, sessionKey)
+
+	// Emulate creation of the obfuscator on node where error have occurred.
+	obfuscator := &OnionObfuscator{
+		sharedSecret: sharedSecrets[len(errorPath)-1],
```

Comment:
The test should also test the case of the error originating from a node that isn't the last hop. I think this would've exposed the issue I pointed out above.